src/lib/include/xtgeo/types.hpp

```diff
@@ -0,0 +1,337 @@
+        }
+        if (actnumsv.shape(0) != ncol || actnumsv.shape(1) != nrow ||
+            actnumsv.shape(2) != nlay) {
+            throw std::runtime_error("actnumsv should have shape (ncol, nrow, nlay)");
+        }
+    };
+};
+
+/*
+ * Need Grid3D cell corners for a single Grid cell for some operations
+ * The cell is a general hexahedron with 8 corners in space, organized as this
+ * for "upper" (top) and "lower" (base). Depth to "upper" is <= depth to "lower"
+ *
+ *  nw ---- ne    nw refers to North-West corner, ne to North-East, etc.
+ *  |       |
+ *  |       |     Notice order: sw - se - nw - ne
+ *  sw ---- se
+ */
+
+struct CellCorners
+{
+    xyz::Point upper_sw;
+    xyz::Point upper_se;
+    xyz::Point upper_nw;
+    xyz::Point upper_ne;
+    xyz::Point lower_sw;
+    xyz::Point lower_se;
+    xyz::Point lower_nw;
+    xyz::Point lower_ne;
+
+    // Default constructor
+    CellCorners() = default;
+
+    // Constructor that takes a one-dimensional array of 24 elements
+    CellCorners(const std::array<double, 24> &arr) :
+      upper_sw(arr[0], arr[1], arr[2]), upper_se(arr[3], arr[4], arr[5]),
+      upper_nw(arr[6], arr[7], arr[8]), upper_ne(arr[9], arr[10], arr[11]),
+      lower_sw(arr[12], arr[13], arr[14]), lower_se(arr[15], arr[16], arr[17]),
+      lower_nw(arr[18], arr[19], arr[20]), lower_ne(arr[21], arr[22], arr[23])
+    {
+    }
+
+    // arrange the corners in a single array for easier access in some cases
+    std::array<double, 24> arrange_corners() const
+    {
+        return {
+            upper_sw.x, upper_sw.y, upper_sw.z, upper_se.x, upper_se.y, upper_se.z,
+            upper_nw.x, upper_nw.y, upper_nw.z, upper_ne.x, upper_ne.y, upper_ne.z,
+            lower_sw.x, lower_sw.y, lower_sw.z, lower_se.x, lower_se.y, lower_se.z,
+            lower_nw.x, lower_nw.y, lower_nw.z, lower_ne.x, lower_ne.y, lower_ne.z
+        };
+    }
+
+    // convert CellCorners struct to a std::array<double, 24> array
+
+    py::array_t<double> to_numpy() const
+    {
+        auto result = py::array_t<double>({ 8, 3 });
+        auto result_m = result.mutable_unchecked<2>();
+        auto corners = arrange_corners();
+        for (size_t i = 0; i < 8; i++) {
+            result_m(i, 0) = corners[3 * i];
+            result_m(i, 1) = corners[3 * i + 1];
+            result_m(i, 2) = corners[3 * i + 2];
+        }
+        return result;
+    }
+};
+}  // namespace grid3d
+
+// =====================================================================================
+
+namespace regsurf {
+
+struct RegularSurface
+{
+    size_t ncol;
+    size_t nrow;
+    double xori;
+    double yori;
+    double xinc;
+    double yinc;
+    double rotation;
+    py::array_t<double> values;
+    py::array_t<bool> mask;
+
+    // Default constructor
+    RegularSurface() = default;
+
+    // Constructor that takes a Python object and a metadata_only flag
+    RegularSurface(const py::object &rs, bool metadata_only = false)
```

Comment:
I'm a bit confused about metadata only, maybe because of thinking so much about dataio... it's meant just to take fields but not the full value-data?

Reply:
I agree it can be confusing. Removed it, as I think the array memory is shared in python and pybind anyway(?)